Sid164 old new public keys

sandJohnson opened the pull request:
Adding functions to Coordination contract. Build fails - CircleCI and Spotify errors. Compile is clean. Don't know how to fix this!!

sandJohnson committed:
Add "Unmasked" to function name to explicitly state the required state of the participant.

sandJohnson committed:
Introduce public key management. In progress

sandJohnson committed:
Getting rid of compile errors - still in progress

sandJohnson committed:
Add functions to handle public key changes

drinkcoffee committed:
Add circle ci to branch

Signed-off-by: Peter  Robinson <drinkcoffee@eml.cc>

drinkcoffee committed:
Move common solidity to new location

Signed-off-by: Peter  Robinson <drinkcoffee@eml.cc>

drinkcoffee committed:
Moved across from master

Signed-off-by: Peter  Robinson <drinkcoffee@eml.cc>

drinkcoffee committed:
Fixed up some comments.

Signed-off-by: Peter  Robinson <drinkcoffee@eml.cc>

drinkcoffee commented on common-solidity/coordination-contract/contracts/CrosschainCoordinationV1.sol:
```diff
@@ -30,6 +30,15 @@ contract CrosschainCoordinationV1 is CrosschainCoordinationInterface {
     uint256 public constant MANAGEMENT_PSEUDO_SIDECHAIN_ID = 0;
 
+    // Public key version number
+    uint private FIRST_PUBLIC_KEY = 0;
```
Should be constant and version numbers should be 64 bit. Hence:

uint64 constant private

drinkcoffee commented on common-solidity/coordination-contract/contracts/CrosschainCoordinationV1.sol:
```diff
@@ -30,6 +30,15 @@ contract CrosschainCoordinationV1 is CrosschainCoordinationInterface {
 
+    // Public key version number
+    uint private FIRST_PUBLIC_KEY = 0;
+    uint [] private publicKeyVersion;
```
Should be uint64[]

sandJohnson replied:
code updated

drinkcoffee commented on common-solidity/coordination-contract/contracts/CrosschainCoordinationV1.sol:
```diff
@@ -30,6 +30,15 @@ contract CrosschainCoordinationV1 is CrosschainCoordinationInterface {
+    uint private FIRST_PUBLIC_KEY = 0;
+    uint [] private publicKeyVersion;
+
+    // Public key & status
```
Are these class / contract level fields used? They aren't inside the sidechain specific structure. As such, I expect they aren't used.

sandJohnson replied:
Some fields have been deleted

drinkcoffee commented on common-solidity/coordination-contract/contracts/CrosschainCoordinationV1.sol:
```diff
@@ -42,6 +51,24 @@ contract CrosschainCoordinationV1 is CrosschainCoordinationInterface {
+    }
+
+    struct PublicKey {
+    //  uint sidechainID;   // Now that I added the struct to sidechains record, we probably don't need this
```
I agree. This should be removed.

sandJohnson replied:
Has been removed

drinkcoffee commented on common-solidity/coordination-contract/contracts/CrosschainCoordinationV1.sol:
```diff
@@ -42,6 +51,24 @@ contract CrosschainCoordinationV1 is CrosschainCoordinationInterface {
+
+    struct PublicKey {
+    //  uint sidechainID;   // Now that I added the struct to sidechains record, we probably don't need this
+        uint versionNumber;
```
uint64

sandJohnson replied:
Done

drinkcoffee commented on common-solidity/coordination-contract/contracts/CrosschainCoordinationV1.sol:
```diff
@@ -42,6 +51,24 @@ contract CrosschainCoordinationV1 is CrosschainCoordinationInterface {
+    struct PublicKey {
+    //  uint sidechainID;   // Now that I added the struct to sidechains record, we probably don't need this
+        uint versionNumber;
+        uint256 status;
```
I think the status should be able to fit into a field smaller than a uint256. If we allow it to be a uint64, then that will allow for a huge variety of status codes, and will combine with the uint64 field above to compress when stored to a single uint256 storage slot.

sandJohnson replied:
Changed to uint64

drinkcoffee commented on common-solidity/coordination-contract/contracts/CrosschainCoordinationV1.sol:
```diff
@@ -42,6 +51,24 @@ contract CrosschainCoordinationV1 is CrosschainCoordinationInterface {
+        uint blockNumber;
+        bytes key;
+    }
+    //mapping (uint => PublicKey) publicKeys;
```
Please remove.

sandJohnson replied:
Done

drinkcoffee commented on common-solidity/coordination-contract/contracts/CrosschainCoordinationV1.sol:
```diff
@@ -214,12 +249,17 @@ contract CrosschainCoordinationV1 is CrosschainCoordinationInterface {
+            // Set up the detail for the proposed key to be associated with this sidechain
+            uint256 sidechainsNoOfKeys = sidechains[_sidechainId].numberOfKeys;
+            uint keyVersion = sidechains[_sidechainId].publicKeys[sidechainsNoOfKeys-1].versionNumber;      // Mapping starts at 0 not 1
+            sidechains[_sidechainId].publicKeys[sidechainsNoOfKeys].versionNumber = keyVersion + 1;
```
The key version is specified by the nodes when the key is being generated. As such, they key version will be passed into this function. It should be the passed in via the _additionalInfo parameter.

drinkcoffee replied:
Given the key version is passed in, the version number needs to be checked to see if it is greater than the highest version that has ever been active.

drinkcoffee commented on common-solidity/coordination-contract/contracts/CrosschainCoordinationV1.sol:
```diff
@@ -214,12 +249,17 @@ contract CrosschainCoordinationV1 is CrosschainCoordinationInterface {
+            uint keyVersion = sidechains[_sidechainId].publicKeys[sidechainsNoOfKeys-1].versionNumber;      // Mapping starts at 0 not 1
+            sidechains[_sidechainId].publicKeys[sidechainsNoOfKeys].versionNumber = keyVersion + 1;
+            sidechains[_sidechainId].publicKeys[sidechainsNoOfKeys].status = uint256(PublicKeyStatus.KEY_PROPOSED);
+            sidechains[_sidechainId].publicKeys[sidechainsNoOfKeys].blockNumber = block.number;
```
To match how the other voting happens, the information about the key that is being voted on should not be stored in the sidechains[_sidechainId].publicKeys[] data structure. It should be stored in the voting structure.

sandJohnson replied:
I have not implemented this, but can discuss later

drinkcoffee commented on common-solidity/coordination-contract/contracts/CrosschainCoordinationV1.sol:
```diff
@@ -214,12 +249,17 @@ contract CrosschainCoordinationV1 is CrosschainCoordinationInterface {
+            uint keyVersion = sidechains[_sidechainId].publicKeys[sidechainsNoOfKeys-1].versionNumber;      // Mapping starts at 0 not 1
+            sidechains[_sidechainId].publicKeys[sidechainsNoOfKeys].versionNumber = keyVersion + 1;
+            sidechains[_sidechainId].publicKeys[sidechainsNoOfKeys].status = uint256(PublicKeyStatus.KEY_PROPOSED);
+            sidechains[_sidechainId].publicKeys[sidechainsNoOfKeys].blockNumber = block.number;
```
I think the block number should be the block number when the key becomes active.

sandJohnson replied:
Agreed

drinkcoffee commented on common-solidity/coordination-contract/contracts/CrosschainCoordinationV1.sol:
```diff
@@ -473,11 +508,148 @@ contract CrosschainCoordinationV1 is CrosschainCoordinationInterface {
+    }
+
+    /**
+        * There was a majority vote to change the public key of the Sidechain, so change the public key as has been voted on
```
The voting algorithm may not have been a majority vote. At this point, all we know is that the voting algorithm indicates the vote succeeded and should be actioned.

sandJohnson replied:
Deleted the word 'majority'

drinkcoffee commented on common-solidity/coordination-contract/contracts/CrosschainCoordinationV1.sol:
```diff
@@ -473,11 +508,148 @@ contract CrosschainCoordinationV1 is CrosschainCoordinationInterface {
+     * @param _sidechainId The 256 bit sidechain identifier to which this public key belongs
+     * @return _activeKey latest active public key
+     */
+    function getLastActivePublicKey(uint256 _sidechainId) private view returns ( uint _versionNumber, uint256 _status, uint _blockNumber, bytes memory _key){
```
If only keys that are successfully voted for are stored in the sidechains[_sidechainId].publicKeys data structure, then the latest public key information will be available at: sidechains[_sidechainId].publicKeys[sidechains[_sidechainId].numberOfKeys]

sandJohnson replied:
We need to talk about this

sandJohnson committed:
Added functions to retrieve the public key for a specific version

Also updated the empty test since getPublicKey now returns more information about the key, added more functions and revised the changePublicKey function

sandJohnson committed:
Deleted 'majority' from the comments

As requested by review of code

drinkcoffee committed:
Merge branch 'master' into sid164-old-new-public-keys
